subsys/testsuite/ztest/src/ztest_new.c

```diff
@@ -356,13 +388,15 @@ static int run_test(struct ztest_suite_node *suite, struct ztest_unit_test *test
 	}
 	run_test_functions(suite, test, data);
 out:
-	ret |= cleanup_test(test);
-	if (test_result != ZTEST_RESULT_SUITE_FAIL) {
+	phase = TEST_PHASE_AFTER;
+	if (test_result != ZTEST_RESULT_SUITE_FAIL && !executed_after_fn) {
+		executed_after_fn = true;
```

Comment:
What you have looks fine, but I have mixed feelings about mixing `goto` and `if(bool)` checks to skip over behavior in the same function.  Could we either create a `out_after` label or possibly clean-up the `setjmp`s and have one jump point, something like 
```
z_test_[pass|skip|fail]() 
{
  longjmp(test_asserted, TC_[PASS|SKIP|FAIL]); // change TC_PASS so != 0
}
:
int run_test() 
{
  ret = setjmp(test_asserted); // Returns 0 in original code. returns TC_[PASS|SKIP|FAIL] from longjmp
  if(ret == 0) { 
    rules(before);
    before(); 
    test_func();
    ret = TC_PASS;
  }
  if(phase != AFTER) {
    phase = AFTER;
    after();
    rules(after);
  }
}
```

Reply:
Turns out this wasn't even needed. It was actually caused by a bug I caught later in my code that set the jump to `test_fail` instead of `stack_fail`.